File: scratch/python/analyse.py
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta, timezone

def get_latest_wind_components(xml_text: str) -> dict[str, str | None]:
    if not xml_text:
        return {"U": None, "V": None}

    ns = {
        "wcs": "http://www.opengis.net/wcs/2.0",
        "ows": "http://www.opengis.net/ows/2.0",
    }

    try:
        root = ET.fromstring(xml_text)
    except ET.ParseError:
        return {"U": None, "V": None}

    u_latest: tuple[datetime | None, str | None] = (None, None)
    v_latest: tuple[datetime | None, str | None] = (None, None)

    for cs in root.findall(".//wcs:CoverageSummary", ns):
        if cs is None:
            continue

        cov_id_el = cs.find("wcs:CoverageId", ns)
        if cov_id_el is None:
            continue

        cov_id = (cov_id_el.text or "").strip()
        if not cov_id:
            continue

        # extraction date sûre
        parts = cov_id.split("___")
        if not parts:
            continue

        date_str = parts[-1]
        try:
            dt = datetime.strptime(date_str, "%Y-%m-%dT%H.%M.%SZ")
        except Exception:
            continue

        if cov_id.startswith("U_COMPONENT_OF_WIND__SPECIFIC_HEIGHT_LEVEL_ABOVE_GROUND___"):
            if u_latest[0] is None or dt > u_latest[0]:
                u_latest = (dt, cov_id)

        elif cov_id.startswith("V_COMPONENT_OF_WIND__SPECIFIC_HEIGHT_LEVEL_ABOVE_GROUND___"):
            if v_latest[0] is None or dt > v_latest[0]:
                v_latest = (dt, cov_id)

    return {
        "U": u_latest[1],
        "V": v_latest[1],
    }


def get_nearest_time(xml_text: str) -> str:
    if not xml_text:
        raise ValueError("xml_text vide")

    ns = {
        "gml": "http://www.opengis.net/gml/3.2",
        "gmlrgrid": "http://www.opengis.net/gml/3.3/rgrid",
    }

    try:
        root = ET.fromstring(xml_text)
    except ET.ParseError as e:
        raise ValueError(f"XML invalide: {e}")

    # beginPosition
    begin_el = root.find(".//gml:beginPosition", ns)
    if begin_el is None or not begin_el.text:
        raise ValueError("beginPosition manquant")

    try:
        base_time = datetime.fromisoformat(begin_el.text.replace("Z", "+00:00"))
    except Exception as e:
        raise ValueError(f"beginPosition invalide: {e}")

    # coefficients time
    coeffs: list[int] | None = None

    for axis in root.findall(".//gmlrgrid:GeneralGridAxis", ns):
        if axis is None:
            continue

        axis_name_el = axis.find("gmlrgrid:gridAxesSpanned", ns)
        if axis_name_el is None or not axis_name_el.text:
            continue

        if axis_name_el.text.strip() == "time":
            coeffs_el = axis.find("gmlrgrid:coefficients", ns)
            if coeffs_el is None or not coeffs_el.text:
                continue

            try:
                coeffs = [int(x) for x in coeffs_el.text.split() if x.strip()]
            except Exception:
                raise ValueError("coefficients time invalides")

            break

    if not coeffs:
        raise ValueError("aucun coefficient temporel trouvé")

    now = datetime.now(timezone.utc)

    # sélection robuste
    def compute_dt(sec: int) -> datetime:
        return base_time + timedelta(seconds=sec)

    best = min(
        coeffs,
        key=lambda s: abs(compute_dt(s) - now)
    )

    exact_time = compute_dt(best)

    # format strict WCS
    return exact_time.strftime("%Y-%m-%dT%H:%M:%SZ")

File: src/lib/wind/analyse.ts
import { XMLParser } from "fast-xml-parser";

interface WindComponents {
  U: string | null;
  V: string | null;
}

export function getLatestWindComponents(xmlText: string): WindComponents {
  if (!xmlText) return { U: null, V: null };

  const parser = new XMLParser({ ignoreAttributes: false, attributeNamePrefix: "@_" });
  let root: any;
  try {
    root = parser.parse(xmlText);
  } catch {
    return { U: null, V: null };
  }

  // Naviguer jusqu'aux CoverageSummary (structure wcs:Capabilities > wcs:Contents > wcs:CoverageSummary)
  const contents =
    root?.["wcs:Capabilities"]?.["wcs:Contents"] ??
    root?.["Capabilities"]?.["Contents"] ??
    root?.["wcs:CoverageDescriptions"] ??
    {};

  const rawSummaries =
    contents?.["wcs:CoverageSummary"] ??
    contents?.["CoverageSummary"] ??
    [];

  const summaries: any[] = Array.isArray(rawSummaries) ? rawSummaries : [rawSummaries];

  let uLatest: [Date | null, string | null] = [null, null];
  let vLatest: [Date | null, string | null] = [null, null];

  for (const cs of summaries) {
    if (!cs) continue;

    const covId: string = (
      cs?.["wcs:CoverageId"] ??
      cs?.["CoverageId"] ??
      ""
    ).trim();

    if (!covId) continue;

    const parts = covId.split("___");
    if (!parts.length) continue;

    const dateStr = parts[parts.length - 1];
    const dt = parseCoverageDate(dateStr);
    if (!dt) continue;

    if (covId.startsWith("U_COMPONENT_OF_WIND__SPECIFIC_HEIGHT_LEVEL_ABOVE_GROUND___")) {
      if (!uLatest[0] || dt > uLatest[0]) uLatest = [dt, covId];
    } else if (covId.startsWith("V_COMPONENT_OF_WIND__SPECIFIC_HEIGHT_LEVEL_ABOVE_GROUND___")) {
      if (!vLatest[0] || dt > vLatest[0]) vLatest = [dt, covId];
    }
  }

  return { U: uLatest[1], V: vLatest[1] };
}

function parseCoverageDate(dateStr: string): Date | null {
  // Format: 2024-01-15T12.00.00Z  (points au lieu de ":" pour HH.MM.SS)
  const normalized = dateStr.replace(
    /(\d{4}-\d{2}-\d{2}T)(\d{2})\.(\d{2})\.(\d{2})Z/,
    "$1$2:$3:$4Z"
  );
  const dt = new Date(normalized);
  return isNaN(dt.getTime()) ? null : dt;
}

// ---------------------------------------------------------------------------

export function getNearestTime(xmlText: string): string {
  if (!xmlText) throw new Error("xmlText vide");

  const parser = new XMLParser({ ignoreAttributes: false, attributeNamePrefix: "@_" });
  let root: any;
  try {
    root = parser.parse(xmlText);
  } catch (e) {
    throw new Error(`XML invalide: ${e}`);
  }

  const rawBegin = findDeep(root, "gml:beginPosition") ?? findDeep(root, "beginPosition");
  if (!rawBegin) throw new Error("beginPosition manquant");

  // Gérer le cas où fast-xml-parser retourne un objet avec attributs
  const beginText = typeof rawBegin === "object"
    ? (rawBegin as any)["#text"]
    : rawBegin;

  if (!beginText) throw new Error("beginPosition manquant");

  const baseTime = new Date(beginText.replace("Z", "+00:00") ?? beginText);
  if (isNaN(baseTime.getTime())) throw new Error(`beginPosition invalide: ${beginText}`);

  // Coefficients de l'axe time
  const axes =
    findAllDeep(root, "gmlrgrid:GeneralGridAxis") ??
    findAllDeep(root, "GeneralGridAxis") ??
    [];

  let coeffs: number[] | null = null;

  for (const axis of axes) {
    const axisName: string = (
      axis?.["gmlrgrid:gridAxesSpanned"] ??
      axis?.["gridAxesSpanned"] ??
      ""
    ).trim();

    if (axisName !== "time") continue;

    const coeffsText: string = (
      axis?.["gmlrgrid:coefficients"] ??
      axis?.["coefficients"] ??
      ""
    ).trim();

    if (!coeffsText) continue;

    try {
      coeffs = coeffsText.split(/\s+/).filter(Boolean).map(Number);
      if (coeffs.some(isNaN)) throw new Error();
    } catch {
      throw new Error("coefficients time invalides");
    }
    break;
  }

  if (!coeffs || coeffs.length === 0) throw new Error("aucun coefficient temporel trouvé");

  const now = Date.now();

  const computeDt = (sec: number): Date => new Date(baseTime.getTime() + sec * 1000);

  const best = coeffs.reduce((a, b) =>
    Math.abs(computeDt(a).getTime() - now) <= Math.abs(computeDt(b).getTime() - now) ? a : b
  );

  return formatWCS(computeDt(best));
}

function formatWCS(dt: Date): string {
  return dt.toISOString().replace(/\.\d{3}Z$/, "Z");
}

// Helpers pour traverser un objet JSON parsé à la recherche d'une clé
function findDeep(obj: any, key: string): any {
  if (!obj || typeof obj !== "object") return null;
  if (key in obj) return obj[key]; // plus de String() cast
  for (const v of Object.values(obj)) {
    const found = findDeep(v, key);
    if (found !== null) return found;
  }
  return null;
}

function findAllDeep(obj: any, key: string): any[] {
  const results: any[] = [];
  if (!obj || typeof obj !== "object") return results;
  if (key in obj) {
    const val = obj[key];
    results.push(...(Array.isArray(val) ? val : [val]));
  }
  for (const v of Object.values(obj)) {
    results.push(...findAllDeep(v, key));
  }
  return results;
}